sympy/core/equation.py

```diff
@@ -0,0 +1,342 @@
+variable. Thus the statement ``Equation/b`` yields a new equation
+``Equation.lhs/b = Equation.rhs/b``
+
+The intent is to allow using the mathematical tools in SymPy to rearrange
+equations and perform algebra in a stepwise fashion. In this way more people
+can successfully perform algebraic rearrangements without stumbling over
+missed details such as a negative sign. This mimics the capabilities available
+in [SageMath](https://www.sagemath.org/) and
+[Maxima](http://maxima.sourceforge.net/).
+"""
+
+from .basic import Basic
+from .sympify import _sympify
+
+
+class Equation(Basic):
+    """
+    This class defines an equation with a left-hand-side (lhs) and a right-
+    hand-side (rhs) connected by the "=" operator (e.g. $p*V = n*R*T$).
+
+    Explanation
+    ===========
+    This class defines relations that all high school and college students
+    would recognize as mathematical equations. At present only the "=" relation
+    operator is recognized.
+
+    This class is intended to allow using the mathematical tools in SymPy to
+    rearrange equations and perform algebra in a stepwise fashion. In this
+    way more people can successfully perform algebraic rearrangements without
+    stumbling over missed details such as a negative sign.
+
+    Create an equation with the call ``Equation(lhs,rhs)``, where ``lhs`` and
+    ``rhs`` are any valid Sympy expression. ``Eqn(...)`` is a synonym for
+    ``Equation(...)``.
+
+    Examples
+    ========
+    >>> from sympy import var, Equation, Eqn, exp, log, integrate, Integral
+    >>> from sympy import solve
+    >>> a, b, c = var('a b c')
+    >>> str(Equation(a,b/c))
+    a=b/c
+    >>> t=Eqn(a,b/c)
+    >>> str(t)
+    a=b/c
+    >>> str(t*c)
+    a*c=b
+    >>> str(c*t)
+    a*c=b
+    >>> str(exp(t))
+    exp(a)=exp(b/c)
+    >>> str(exp(log(t)))
+    a=b/c
+
+    Integration can only be performed on one side at a time.
+    >>> q=Eqn(a*c,b/c)
+    >>> integrate(q,b,side='rhs')
+    b**2/(2*c)
+    >>> integrate(q,b,side='lhs')
+    a*b*c
+
+    Make a pretty statement of integration from an equation
+    >>> str(Eqn(Integral(q.lhs,b),integrate(q,b,side='rhs')))
+    Integral(a*c, b)=b**2/(2*c)
+
+    This is duplicated by the convenience function self.integ
+    >>> str(q.integ(b))
+    Integral(a*c, b)=b**2/(2*c)
+
+    SymPy's solvers do not understand these equations. They expect an
+    expression that the solver assumes = 0. Thus to use the solver the
+    equation must be rearranged so that all non-zero symbols are on one side.
+    Then just the non-zero symbolic side is passed to ``solve()``.
+    >>> t2 = t-t.rhs
+    >>> str(t2)
+    a-b/c=0
+    >>> solve(t2.lhs,c)
+    [b/a]
+    """
+
+    def __new__(cls, lhs, rhs, **kwargs):
+        check = kwargs.pop('check', True)
+        lhs = _sympify(lhs)
+        rhs = _sympify(rhs)
+        if check:
+            lsimp = lhs.simplify()
+            rsimp = rhs.simplify()
```

Comment:
The `simplify` function can be an enormously expensive routine. Building it into the `Equation` class by default like this would make using this class a non-starter in all but the most basic situations. It also ruins various possibilities like the example I showed above since
```julia
In [3]: (cos(x) + I*sin(x)).simplify()                                                                                                                        
Out[3]: 
 ⅈ⋅x
ℯ  
```
And there is no way to disable this simplification when using the operators like `eq1 + eq2`.

Reply:
I'm confused. Did I set this up wrong? My intent was to only do the simplification as part of the test. It should not change lhs and rhs.

Reply:
I do not think I understand what your are getting at.
```
>>> Eqn(cos(a)+I*sin(a),exp(I*a))
I*sin(a) + cos(a) = exp(I*a)
```
and
```
>>> Eqn(cos(a)+I*sin(a),exp(I*a)).simplify()
exp(I*a) = exp(I*a)
```
That said, I do think I am missing something about when expressions automatically simplify. Even when I force `check=False`, so that the part of the code that calls `.simplify()` is skipped, I see simplification of expressions that evaluate to numbers.
```
>>> Eqn(cos(pi)+I*sin(pi),exp(I*pi), check=False)
-1 = -1
```
Likewise, completely avoiding the `Eqn` class:
```
>>> cos(pi)+I*sin(pi)
-1
```
So is the suggestion that a change is needed in `Expr` or some kind of override of the behavior of `Expr`? Or am I completely misunderstanding the issue?

Reply:
Sorry I misunderstood. You're right that this won't change the lhs and rhs as they appear in the resulting equation.

I still don't think that we can use simplify here though because of the computational cost. This class will be unusable in many situations if it calls simplify at all let alone at each stage of manipulation.

Reply:
If you look at all the internal calls to the class `Eqn` you will see that they are made with `check=False` (see lines: 148, 154,157, 161, 164, 169, 172, 289, 212, 215, 276, 283, 288, 296, 303,  317, and 328). Thus only a call made directly from the command line without `check=False` will result in the call to simplify. Now, I suppose there is the issue of somebody coding a use of `Eqn` that produces something difficult to simplify and forgetting to use `check=False`, but I think the most common use for `Eqn` will be for interactive manipulation. I have trouble imagining a situation where the manual input into of an `Eqn` would be significantly burdened by this check.

Reply:
I guarantee that there will be bug reports complaining about this. The manual input can be still from the output of something else:
```julia
In [3]: a, b, c, d, e, x = symbols('a, b, c, d, e, x')                                                                            

In [4]: r1, r2, r3, r4 = roots(a*x**4 + b*x**3 + c*x**2 + d*x + e, x)                                                             

In [5]: %time eq = Eqn(x, r1)
```
Note that in this example `is_number` is False on both sides of the equation but simplify is called first.

Also in general `simplify(a) != simplify(b)` can not conclude that the two are unequal. If both are `is_number` then `(a-b).evalf()` can but that can also be expensive e.g. `(exp(exp(exp(100)))-1).evalf()`.

It wouldn't be hard to construct an example without symbols e.g. `roots(x**4 + sqrt(2)*x**3 + I*x**2 + 27*x + 4, x)`
but I doubt that many users would want to use an equation without symbols on either side anyway so it hardly seems worth it. Even checking `is_number` can be slow with large expressions.

If you really wanted to do this then a much lighter-weight check would be `Eq(lhs, rhs) is S.false`. I don't think that we should even do that by default though.

Reply:
> If you look at all the internal calls to the class `Eqn` you will see that they are made with `check=False`

Sorry I didn't see that.

Reply:
I've been puttering with this while trying to get a large quantum package to recompile and pass all its tests.

`exp(exp(exp(100)))-1)` causes C integer overflow problems on my installation. However, I could get tests running with `exp(exp(exp(20)))-1`. I find that calls like `str(exp(exp(exp(20)))-1)` take a very long time. Also calls to `Basic(exp(exp(exp(20)))-1,1)` take a very long time. The `Eqn` code is only slow when it must make either calls to `str` or `Basic`.

If I back down to `exp(exp(exp(10)))-1` everything works in the 1 - 2 ms time frame on my laptop. I think the evaluation built into `str` and somewhere in `Basic` must be on the edge of stability as the limit on C integer sizes are approached.

If I need a break from my compilation odyssey I will look more into what happens in `class Basic`.

Reply:
The cases you describe are all bugs related to the presumption that it is safe to call `.evalf()` in library code. That's why I wouldn't want to do it in the `Equation` class. I would also like to remove those other evalf calls and make evalf safer for large inputs but neither of those is an easy thing to do.

Note that `Eq(a, b)` does indeed do `(a-b).evalf()` which is a problem but there it is for a good reason which is to be able to evaluate to False as the Eq class is often used with things that are likely to be unequal.

Reply:
My tests suggest that in cases where a==b should be true so that most things cancel out, `(a-b).evalf()` is unlikely to cause problems. That said even just the expression `exp(exp(exp(20)))` spins its wheels in my tests.

I wonder if the solution is not about removing `.evalf()` calls, but about getting some intelligent error reporting about stability from some of the evaluation algorithms.

Reply:
The *worst* case scenario for `(a-b).evalf()` is precisely when `a` and `b` *are* equal but are represented in different ways so that they don't cancel out symbolically but they do numerically. There is no way to prove that an expression is numerically equal so evalf just keeps increasing the precision until it hits the limit. If the numerical evaluation of the terms is expensive at high precision then the result can be expensive:
```julia
In [1]: r = Poly(x**5 - x + 1, x).root(1)                                                                                                      

In [2]: r                                                                                                                                      
Out[2]: 
       ⎛ 5           ⎞
CRootOf⎝x  - x + 1, 1⎠

In [3]: a = exp(I*r)                                                                                                                           

In [4]: b = cos(r) + I*sin(r)                                                                                                                  

In [5]: %time (a-b).evalf()                                                                                                                    
CPU times: user 23.5 s, sys: 103 ms, total: 23.6 s
Wall time: 24 s
Out[5]: 0.e-135 + 0.e-135⋅ⅈ
```

> I wonder if the solution is not about removing `.evalf()` calls, but about getting some intelligent error reporting about stability from some of the evaluation algorithms.

The solution is both. In some cases like `exp(exp(...))` the result in floating point has a mantissa that exceeds the memory capacity of any computer so that the calculation can never succeed and should not be attempted in any case. In other cases evalf can work but is slow (as for CRootOf). So yes evalf should have some control over the mantissa which should be limited the same way that the precision is. However also we should avoid calling evalf unnecessarily.

Reply:
Your example definitely illustrates a slow process. However, it really bothered me because I can look at `a-b` and recognize that it is `0` in a few seconds. This suggested that SymPy was not using all the information available to do the simplification. As the _proof-of-principle_ below illustrates, that is indeed the case. The bit of code I've written is far from being generally usable, but it might be a place to start.
```
"""
Proof of principle using hashes to speed up simplification by substituting
symbols for equivalent expressions.
"""
from sympy.core.expr import Expr
from sympy.core.symbol import Symbol
from sympy.core.numbers import *

def presimpstst(expr):
    def makehashtable(expr, hashtable=[]):
        for obj in expr.args:
            if len(hashtable) == 0:
                hashtable.append([obj.__hash__(), obj])
            else:
                found = False
                for k in range(len(hashtable)):
                    if hashtable[k][0] == obj.__hash__():
                        (hashtable[k]).append(obj)
                        found = True
                if not (found):
                    hashtable.append([obj.__hash__(), obj])
            if isinstance(obj, Expr):
                makehashtable(obj, hashtable)
        pass

    def makesublst(expr):
        lst = []
        fnllst = []
        makehashtable(expr, lst)
        for i in range(len(lst)):
            if len(lst[i]) > 2 and not (isinstance(lst[i][1], Integer)
                                        or isinstance(lst[i][1], float)
                                        or isinstance(lst[i][1], Symbol)):
                if lst[i][1] != I:
                    fnllst.append(lst[i])
        return fnllst

    def crawlexprtree(expr, subhash, hashidx, didsubs=False):
        exprtype=type(expr)
        tempargs = list(expr.args)
        for k in range(len(tempargs)):
            if tempargs[k].__hash__() == subhash:
                tempsym = Symbol('sub_'+str(hashidx))
                tempargs[k] = tempsym
                didsubs = True
            if isinstance(tempargs[k],Expr):
                if tempargs[k] != I:
                    if not(isinstance(tempargs[k],Symbol)
                           or isinstance(tempargs[k],float)
                           or isinstance(tempargs[k],Integer)
                           or isinstance(tempargs[k],Number)):
                        tempargs[k], didsubs = crawlexprtree(tempargs[k],
                                                             subhash, hashidx,
                                                             didsubs)
        tempargs = tuple(tempargs)
        newexpr = exprtype(*tempargs)
        return newexpr, didsubs
    sublst = makesublst(expr)
    didsublst = [False]*len(sublst)
    exprtype=type(expr)
    temp = expr
    for k in range(len(sublst)):
        temp,didsublst[k] = crawlexprtree(temp, sublst[k][0], k)
    return temp,didsublst
```
Examples:
```
In [5]: r1=poly(x**6+3*x+7).root(1)                                             

In [6]: %time simplify(presimpstst(cos(r1)+I*sin(r1)-exp(I*r1))[0])             
CPU times: user 596 ms, sys: 8.09 ms, total: 604 ms
Wall time: 603 ms
Out[6]: 0

In [7]: %time simplify(cos(r1)+I*sin(r1)-exp(I*r1))                             
CPU times: user 33.4 s, sys: 8.75 ms, total: 33.4 s
Wall time: 33.4 s
Out[7]: 0

In [8]: %time simplify(presimpstst(cos(x)+I*sin(x)-exp(I*x))[0])                
CPU times: user 155 ms, sys: 14 µs, total: 155 ms
Wall time: 154 ms
Out[8]: 0

In [9]: %time simplify(cos(x)+I*sin(x)-exp(I*x))                                
CPU times: user 138 ms, sys: 0 ns, total: 138 ms
Wall time: 137 ms
Out[9]: 0

```
So, the pre-simplification substitution adds a little overhead time (see `[8]` and `[9]`), but for expressions with repeated complicated sub-expressions it can save a lot of time (see `[6]` versus `[7]`), if things cancel out. 

To be usable there would also have to be back-substitution after the simplification. I also think the tests to decide what should not be substituted are not complete.

Reply:
When looking at particular expressions it is usually possible to come up with a fast way of simplifying. The hardest thing is writing a generic `simplify` routine that should be expected to work efficiently for *all* possible inputs. That's why the use of `simplfy` is strongly discouraged in library code. Improvements to the generic `simplify` function are of course welcome but they won't make it reasonable to call simplify routinely.

Reply:
@oscarbenjamin I agree, fast simplification for all expressions seems unlikely. My familiarity with the SymPy codebase, although improving, is still rather superficial. Do you or others have any sense of whether hashes are done well enough in the code that they can be reliably used to identify equivalent parts of expressions, as I've done in my [example above](https://github.com/sympy/sympy/pull/19479#discussion_r437429772)? Does this appear to be worth pursuing as a separate issue?

From my perspective as an end user, 30 seconds to wait for sanity checking that saves me minutes trying to figure out where I have a typo is worth a lot. The cutoff for being useful in saving time is actually around 5 minutes in my experience. One solution (that I see has had some discussion  #9691) would be to build a timeout that raises an error  into potentially long running things such as `simplify`. Set the default to what people think is a reasonable waiting time and allow calls to long running functions to set the `timeout` value to shorter or longer than that. In #9691, I see people discussed using some of the other timing tools, but I think doing that well will require spawning sympy operations on a separate thread. Recreating the environment properly on the launched thread might be difficult.

Reply:
> Do you or others have any sense of whether hashes are done well enough in the code that they can be reliably used to identify equivalent parts of expressions

I'm not sure I've fully followed the code. For the most part sympy's hashes work correctly which means that if `a == b` then `hash(a) == hash(b)`. It's not clear to me whether the code you've shown relies in the (not guaranteed) converse `hash(a)==hash(b) -> a==b`.

Reply:
That's exactly what I was worried about. Presently, the code simply assumes `hash(a)==hash(b) -> a==b`. If that isn't true than further checking is necessary when `hash(a)==hash(b)`. As the code already walks the whole of the expression tree, that probably will not change the efficiency a lot. However, it would need some thought. I'm not sure I presently have time to add this to my projects. 

I will try to write up an explanation of the idea and add it to issues. Maybe somebody else has some time and is interested in improving simplification efficiency.

Reply:
Ideas summarized in #19525